hooks/pre_validations/05-set_hostname.rb

```diff
@@ -0,0 +1,43 @@
+unless app_value(:hostname).nil?
+  hostname = app_value(:hostname)
+
+  command = "hostname \"#{hostname}\" 2>&1"
+  logger.debug(command)
+  unless kafo.noop?
+    result = `command`
+    if $?.exitstatus > 0
+      say "<%= color('Hostname setup failed: #{result}', :bad) %>"
+      kafo.class.exit(101)
+    end
+  end
+
+  filename = '/etc/sysconfig/network'
+  logger.debug("Writing hostname (#{hostname}) to #{filename}")
+  unless kafo.noop?
+    entry = "HOSTNAME=#{hostname}"
+    begin
+      text = File.read(filename) 
+      text.gsub!(/^HOSTNAME=.*?$/, entry)
+      text += "#{entry}\n" unless text =~ /^HOSTNAME=/
+      File.open(filename, "w") { |file| file.write(text) }
+    rescue => error
+      say "<%= color('Hostname setup failed: #{error}', :bad) %>"
+      kafo.class.exit(101)
+    end
+  end
+
+  # set /etc/hosts if needed
+  `hostname -f`
+  if $?.exitstatus > 0 
+    filename = '/etc/hosts'
+    logger.debug("hostname -f failed. Setting temporary hostname for loopback in #{filename}")
+    unless kafo.noop?
+      open(filename, 'a') do |f|
+        f << "# temporary hostname for loopback\n"
+        f << "127.0.0.1 #{hostname}\n"
+      end
+    end
+  end
```

Comment:
unfortunately puppets host type is pretty bad.

If you have a pre existing entry for an ip address it will just add another entry, which is wrong.

There is the possibility to purge the file before rewriting it using puppet, but entries are keyed on name, so restoring the '127.0.0.1 localhost...' and '::1 localhost...' entries becomes impossible using puppet.

My thought, as to the best we can do in the event that we're going to set up the hostname is replace /etc/hosts with just the lines
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4
::1	  localhost localhost.localdomain localhost6 localhost6.localdomain6

and then letting the fusor_network module set up the entry for whichever interface/address they're going to use. 

This is of course at this risk of nuking any existing hosts entries they have, but using the host resource purge in puppet would be just as bad.